File: app/src/main/java/com/door43/translationstudio/ui/translate/FirstTabFragment.java
package com.door43.translationstudio.ui.translate;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;

import org.unfoldingword.door43client.models.Translation;
import org.unfoldingword.resourcecontainer.Project;
import org.unfoldingword.tools.logger.Logger;

import com.door43.translationstudio.core.TargetTranslation;
import com.door43.translationstudio.databinding.FragmentFirstTabBinding;
import com.door43.translationstudio.ui.BaseFragment;
import com.door43.translationstudio.ui.viewmodels.TargetTranslationViewModel;

import org.json.JSONException;

import java.util.List;

/**
 * Gives some instructions when no source text has been selected
 */
public class FirstTabFragment extends BaseFragment implements ChooseSourceTranslationDialog.OnClickListener {

    private OnEventListener listener;
    protected TargetTranslationViewModel viewModel;

    private FragmentFirstTabBinding binding;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentFirstTabBinding.inflate(inflater, container, false);
        viewModel = new ViewModelProvider(requireActivity()).get(TargetTranslationViewModel.class);

        Bundle args = getArguments();
        assert args != null;

        setupObservers();

        try {
            Project p = viewModel.getProject();
            binding.sourceTranslationTitle.setText(
                    p.name + " - " + viewModel.getTargetTranslation().getTargetLanguageName()
            );
        } catch (Exception e) {
            Logger.e(
                FirstTabFragment.class.getSimpleName(),
                "Error getting resource container for '" + viewModel.getTargetTranslation().getId() + "'",
                e
            );
        }

        View.OnClickListener clickListener = v -> {
            FragmentTransaction ft = getParentFragmentManager().beginTransaction();
            Fragment prev = getParentFragmentManager().findFragmentByTag("tabsDialog");
            if (prev != null) {
                ft.remove(prev);
            }
            ft.addToBackStack(null);

            ChooseSourceTranslationDialog dialog = new ChooseSourceTranslationDialog();
            Bundle args1 = new Bundle();
            args1.putString(ChooseSourceTranslationDialog.ARG_TARGET_TRANSLATION_ID, viewModel.getTargetTranslation().getId());
            dialog.setOnClickListener(FirstTabFragment.this);
            dialog.setArguments(args1);
            dialog.show(ft, "tabsDialog");
        };

        binding.newTabButton.setOnClickListener(clickListener);
        binding.secondaryNewTabButton.setOnClickListener(clickListener);

        // attach to tabs dialog
        if(savedInstanceState != null) {
            ChooseSourceTranslationDialog dialog = (ChooseSourceTranslationDialog) getParentFragmentManager().findFragmentByTag("tabsDialog");
            if(dialog != null) {
                dialog.setOnClickListener(this);
            }
        }

        return binding.getRoot();
    }

    private void setupObservers() {
        viewModel.getListItems().observe(getViewLifecycleOwner(), items -> {
            if (!items.isEmpty() && listener != null) {
                listener.onHasSourceTranslations();
            }
        });
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            this.listener = (OnEventListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context + " must implement FirstTabFragment.OnEventListener");
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    /**
     * user has selected to update sources
     */
    public void onUpdateSources() {
        if(listener != null) listener.onUpdateSources();
    }

    @Override
    public void onCancelTabsDialog(String targetTranslationId) {

    }

    @Override
    public void onConfirmTabsDialog(@NonNull List<String> sourceTranslationIds) {
        String[] oldSourceTranslationIds = viewModel.getOpenSourceTranslations();
        for(String id:oldSourceTranslationIds) {
            viewModel.removeOpenSourceTranslation(id);
        }

        if(!sourceTranslationIds.isEmpty()) {
            // save open source language tabs
            for(String slug:sourceTranslationIds) {
                Translation t = viewModel.getTranslation(slug);
                int modifiedAt = viewModel.getResourceContainerLastModified(t);
                try {
                    viewModel.addOpenSourceTranslation(slug);
                    TargetTranslation targetTranslation = viewModel.getTargetTranslation();
                    try {
                        targetTranslation.addSourceTranslation(t, modifiedAt);
                    } catch (JSONException e) {
                        Logger.e(this.getClass().getName(), "Failed to record source translation (" + slug + ") usage in the target translation " + targetTranslation.getId(), e);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }

            // redirect back to previous mode
            if(listener != null) listener.onHasSourceTranslations();
        }
    }

    public interface OnEventListener {
        void onHasSourceTranslations();

        /**
         * user has selected to update sources
         */
        void onUpdateSources();
    }
}

File: app/src/main/java/com/door43/translationstudio/ui/translate/FirstTabFragment.kt
package com.door43.translationstudio.ui.translate

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.door43.translationstudio.databinding.FragmentFirstTabBinding
import com.door43.translationstudio.ui.BaseFragment
import com.door43.translationstudio.ui.viewmodels.TargetTranslationViewModel
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import org.json.JSONException
import org.koin.androidx.viewmodel.ext.android.activityViewModel
import org.unfoldingword.tools.logger.Logger

/**
 * Gives some instructions when no source text has been selected
 */
class FirstTabFragment : BaseFragment(), ChooseSourceTranslationDialog.OnClickListener {

    private var listener: OnEventListener? = null

    private val viewModel: TargetTranslationViewModel by activityViewModel()

    private var _binding: FragmentFirstTabBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentFirstTabBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val args = arguments
        requireNotNull(args)

        setupObservers()

        try {
            val p = viewModel.getProject()
            binding.sourceTranslationTitle.text = "${p?.name} - ${viewModel.targetTranslation.targetLanguageName}"
        } catch (e: Exception) {
            Logger.e(
                FirstTabFragment::class.java.simpleName,
                "Error getting resource container for '${viewModel.targetTranslation.id}'",
                e
            )
        }

        val clickListener = View.OnClickListener {
            val ft = parentFragmentManager.beginTransaction()
            val prev = parentFragmentManager.findFragmentByTag("tabsDialog")
            if (prev != null) {
                ft.remove(prev)
            }
            ft.addToBackStack(null)

            val dialog = ChooseSourceTranslationDialog()
            val args1 = Bundle()
            args1.putString(
                ChooseSourceTranslationDialog.ARG_TARGET_TRANSLATION_ID,
                viewModel.targetTranslation.id
            )
            dialog.setOnClickListener(this@FirstTabFragment)
            dialog.arguments = args1
            dialog.show(ft, "tabsDialog")
        }

        binding.newTabButton.setOnClickListener(clickListener)
        binding.secondaryNewTabButton.setOnClickListener(clickListener)

        // attach to tabs dialog
        if (savedInstanceState != null) {
            val dialog = parentFragmentManager.findFragmentByTag("tabsDialog") as? ChooseSourceTranslationDialog
            dialog?.setOnClickListener(this)
        }
    }

    private fun setupObservers() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    viewModel.model
                        .map { it.items }
                        .distinctUntilChanged()
                        .collect { items ->
                            if (items.isNotEmpty()) {
                                listener?.onHasSourceTranslations()
                            }
                        }
                }
            }
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            this.listener = context as OnEventListener
        } catch (e: ClassCastException) {
            throw ClassCastException("$context must implement FirstTabFragment.OnEventListener")
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    /**
     * user has selected to update sources
     */
    override fun onUpdateSources() {
        listener?.onUpdateSources()
    }

    override fun onCancelTabsDialog(targetTranslationId: String) {}

    override fun onConfirmTabsDialog(sourceTranslationIds: List<String>) {
        val oldSourceTranslationIds = viewModel.getOpenSourceTranslations()
        for (id in oldSourceTranslationIds) {
            viewModel.removeOpenSourceTranslation(id)
        }

        if (sourceTranslationIds.isNotEmpty()) {
            // save open source language tabs
            for (slug in sourceTranslationIds) {
                val t = viewModel.getTranslation(slug)
                if (t != null) {
                    val modifiedAt = viewModel.getResourceContainerLastModified(t)
                    try {
                        viewModel.addOpenSourceTranslation(slug)
                        val targetTranslation = viewModel.targetTranslation
                        try {
                            targetTranslation.addSourceTranslation(t, modifiedAt)
                        } catch (e: JSONException) {
                            Logger.e(
                                this.javaClass.name,
                                "Failed to record source translation ($slug) usage in the target translation ${targetTranslation.id}",
                                e
                            )
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
            }

            // redirect back to previous mode
            listener?.onHasSourceTranslations()
        }
    }

    interface OnEventListener {
        fun onHasSourceTranslations()

        /**
         * user has selected to update sources
         */
        fun onUpdateSources()
    }
}